Update regression tests to run!

peverwhee opened the pull request:
Tag name (required for release branches): sima0_00_001
Originator(s): peverwhee

Description (include the issue title, and the keyword ['closes', 'fixes', 'resolves'] followed by the issue number):
closes #229 (Need ability for ncdata_check routines to end model run)
addresses #230 (Need CIME smoke tests for FPHYStest compset and various physics suites)

**Updates the regression tests to actually run and compare against previous baselines.**

Describe any changes made to build system: 
- added new compset FCAM7 that exercises the cam7 physics suites
- updated config_pes.xml for ne3 grid

Describe any changes made to the namelist:
- updated namelist defaults for ne3 grid (though these are currently unused; using analytic_ic until the SE dycore is updated)

List any changes to the defaults for the input datasets (e.g. boundary datasets): none

List all files MOVED and why:
cime_config/testdefs/testmods_dirs/cam/outfrq_kessler_mpas_derecho_nooutput/user_nl_cam
cime_config/testdefs/testmods_dirs/cam/outfrq_kessler_mpas_derecho_nooutput/shell_commands
cime_config/testdefs/testmods_dirs/cam/outfrq_held_suarez_derecho_nooutput/shell_commands
cime_config/testdefs/testmods_dirs/cam/outfrq_held_suarez_derecho_nooutput/user_nl_cam
cime_config/testdefs/testmods_dirs/cam/outfrq_kessler_derecho_nooutput/shell_commands
cime_config/testdefs/testmods_dirs/cam/outfrq_kessler_derecho_nooutput/user_nl_cam
cime_config/testdefs/testmods_dirs/cam/outfrq_tj2016_derecho_nooutput/user_nl_cam
cime_config/testdefs/testmods_dirs/cam/outfrq_tj2016_derecho_nooutput/shell_commands
- removed "nooutput" from directory name

cime_config/testdefs/testmods_dirs/cam/outfrq_kessler_se_cslam_derecho_nooutput/user_nl_cam
cime_config/testdefs/testmods_dirs/cam/outfrq_kessler_se_cslam_derecho_nooutput/shell_commands
- moved to outfrq_se_sclam/*

List all files added and what they do:
A cime_config/config_archive.xml
- brought over from CAM so tests will run

List all existing files that have been modified, and describe the changes: 
(Helpful git command: `git diff --name-status development...<your_branch_name>`)
M cime_config/config_component.xml
- added FCAM7 configuration

M cime_config/config_compset.xml
- added FCAM7 compset

M cime_config/config_pes.xml
- modified for ne3 grid

M cime_config/namelist_definition_cam.xml
- added ne3 grid defaults

M cime_config/testdefs/testlist_cam.xml
- added FHS94 and FTJ16 SE CSLAM tests
- modified existing FKESSLER SE CSLAM test to use ne5

M src/dynamics/tests/namelist_definition_analy_ic.xml
- use held-suarez initial conditions by default (not "none")

M src/physics/utils/phys_comp.F90
M src/physics/utils/physics_data.F90
- add namelist variable to endrun on diff found during ncdata_check

M test/system/archive_baseline.sh
- update to add a symlink to the ${baseline_dir}/latest_${CAM_FC} directory for easier test_driver usage
- add flag to turn off symlink addition

M test/system/test_driver.sh
- use ${baseline_dir}/latest_${CAM_FC} by default
- add flag to turn off default usage (no baseline comparison)

If there are new failures (compared to the `test/existing-test-failures.txt` file),
have them OK'd by the gatekeeper, note them here, and add them to the file.
If there are baseline differences, include the test and the reason for the
diff. What is the nature of the change? Roundoff?

derecho/intel/aux_sima:
SMS_Ln2.mpasa480_mpasa480.FKESSLER.derecho_intel.cam-outfrq_kessler_mpas_derecho
- will fail until MPAS is fully integrated

derecho/gnu/aux_sima:
SMS_Ln2.mpasa480_mpasa480.FKESSLER.derecho_gnu.cam-outfrq_kessler_mpas_derecho
- will fail until MPAS is fully integrated

If this changes climate describe any run(s) done to evaluate the new
climate in enough detail that it(they) could be reproduced: n/a

CAM-SIMA date used for the baseline comparison tests if different than latest: first baselines!

nusbaume reviewed (changes requested):
Thanks for getting the tests going!  I do have some change requests (which is mostly me backpedaling from previous things I asked you to do, sorry!), but hopefully it won't be too hard to implement.  Of course please let me know if that is not the case.  Thanks again!

peverwhee commented on cime_config/testdefs/testmods_dirs/cam/outfrq_held_suarez_derecho/user_nl_cam:
```diff
@@ -1,4 +1,7 @@
+ncdata_check=/glade/campaign/cesm/community/amwg/sima_baselines/cam_sima_test_snapshots/cam_ne3pg3_held_suarez_snapshot_derecho_gnu_cam_after_c20240412.nc
 debug_output=0
 pver=30
+hist_add_inst_fields;h2: Q, T
```
Currently, "ZM" is geopotential height (wrt surface) in CAM-SIMA - should this be "Z3" instead?

nusbaume replied:
Whoops, sorry, I meant `ZM`, I just knew it as `Z3` from CAM (but I never liked the name because I have no idea what the "3" means).  Thanks!

peverwhee replied:
changed!

peverwhee commented on cime_config/config_component.xml:
```diff
@@ -120,6 +121,7 @@
     <valid_values></valid_values>
     <default_value></default_value>
     <values match="last" modifier='additive'>
+      <value compset="_CAM70">--physics-suites cam7 --analytic_ic</value>
```
moved & added a new outfrq_se_cslam_analy_ic test case directory

peverwhee commented on cime_config/testdefs/testlist_cam.xml:
```diff
@@ -7,56 +7,88 @@
     <options>
       <option name="wallclock">00:10:00</option>
-      <option name="comment">GNU build test for Moist Held-Suarez physics</option>
+      <option name="comment">Test for Moist Held-Suarez physics</option>
```
updated!

peverwhee commented on cime_config/testdefs/testlist_cam.xml:
```diff
@@ -7,56 +7,88 @@
 
   <!-- Derecho dycore tests -->
-  <test compset="FKESSLER" grid="mpasa480_mpasa480" name="SMS_Ln2" testmods="cam/outfrq_kessler_mpas_derecho_nooutput/">
+  <test compset="FKESSLER" grid="mpasa480_mpasa480" name="SMS_Ln2" testmods="cam/outfrq_kessler_mpas_derecho">
```
updated!

peverwhee replied:
it was! thanks! fixed now.

peverwhee commented on cime_config/testdefs/testlist_cam.xml:
```diff
@@ -7,56 +7,88 @@
+  </test>
+
+  <!-- CAM7 physics tests -->
+  <test compset="FCAM7" grid="ne5pg3_ne5pg3_mg37" name="SMS_Ln2" testmods="cam/outfrq_se_cslam">
```
updated!

peverwhee commented on cime_config/testdefs/testlist_cam.xml:
```diff
@@ -7,56 +7,88 @@
+    </machines>
+    <options>
+      <option name="wallclock">00:10:00</option>
+      <option name="comment">Test for SE-CSLAM dycore (with CAM7 physics)</option>
```
updated!

peverwhee commented on cime_config/testdefs/testmods_dirs/cam/outfrq_kessler_mpas_derecho/user_nl_cam:
```diff
@@ -0,0 +1,4 @@
+debug_output=0
+hist_add_inst_fields;h2: Q, T
```
added fields!

peverwhee commented on cime_config/testdefs/testmods_dirs/cam/outfrq_se_cslam/user_nl_cam:
```diff
@@ -0,0 +1,3 @@
+hist_add_inst_fields;h1: T, Q, PS
```
added fields!

peverwhee commented on src/dynamics/tests/namelist_definition_analy_ic.xml:
```diff
@@ -19,7 +19,7 @@
     </desc>
     <values>
-      <value>none</value>
+      <value>held_suarez_1994</value>
```
switched back to none and updated the cam7 user_nl_cam file

peverwhee commented on test/system/archive_baseline.sh:
```diff
@@ -35,62 +28,84 @@ BASELINE ARCHIVED LOCATION
+SYMLINK
+        By default, this script will create a symlink between the new baseline directory and
+        $baseline_dir/latest_${CAM_FC} so that future tests can be run against these baselines
+        until thenext baselines are established. If you'd like to not create the symlink (e.g.
```
fixed!

peverwhee commented on test/system/archive_baseline.sh:
```diff
@@ -35,62 +28,84 @@ BASELINE ARCHIVED LOCATION
 
 WORK FLOW
 
 	This is an example for izumi.
```
updated!

peverwhee commented on test/system/archive_baseline.sh:
```diff
@@ -35,62 +28,84 @@ BASELINE ARCHIVED LOCATION
-    cam_tag=$1
+    cam_tag=${cam_tag}_${CAM_FC,,}
     baselinedir="/glade/campaign/cesm/community/amwg/sima_baselines/$cam_tag"
+    chmod_cmd=""
```
no reason i can think of - changed!

peverwhee commented on test/system/archive_baseline.sh:
```diff
@@ -110,6 +125,12 @@ if [ -n "$CESM_TESTDIR" ]; then
+      if [ "${symlink}" = true ]; then
+        echo "Establishing symlink from '$root_baselinedir/latest_${CAM_FC,,}' to '$root_baselinedir/$cam_tag'"
+	rm -rf $root_baselinedir/latest_${CAM_FC,,}
+        ln -s $root_baselinedir/$cam_tag $root_baselinedir/latest_${CAM_FC,,}
```
thanks! that does indeed work! changed!

peverwhee commented:
@nusbaume thanks for your review. I also added the NL option to endrun on ncdata_check failures (and added it to the kessler test, which is currently the only one that has very small diffs) - if you could take a look at that, that'd be great!

nusbaume reviewed (changes requested):
Almost there!   Just two last requests.

peverwhee commented on src/data/write_init_files.py:
```diff
@@ -1328,6 +1341,12 @@ def write_phys_check_subroutine(outfile, host_dict, host_vars, host_imports,
+    outfile.comment("Endrun if differences were found on this timestep and err_on_fail=TRUE", 2)
+    outfile.write("if (overall_diff_found .and. err_on_fail .and. masterproc) then", 2)
+    outfile.write("call endrun('ERROR: Difference(s) found during ncdata check', file=__FILE__, line=__LINE__)", 3)
+    outfile.write("end if", 2)
```
Good idea - moved!

nusbaume reviewed (approved):
Looks great to me now.  Thanks again!

nusbaume commented:
Also forgot to mention that you can close issue #229 once this PR is merged.  It might also be worth "addressing" issue #230 as well (we can close that one once we have all of snapshot checks working).  Thanks!

mwaxmonsky reviewed (approved):
Looks great @peverwhee, thanks for getting this added!

One minor whitespace fix I think might be needed but nothing worth holding up this PR.

peverwhee commented on test/system/test_driver.sh:
```diff
@@ -132,6 +134,9 @@ while [ "${1:0:1}" == "-" ]; do
         --no-baseline ) no_baseline=false
              ;;
 
+	--no-bl-compare ) baseline_default=false
```
Can I ask why this whitespace change? I'm not a bash expert, but it does match the rest of the arguments

peverwhee replied:
turns out it was a "tab" whitespace - fixed!
